harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -28,15 +34,138 @@
     pass
 
 
-class PyTorchTrialController(det.TrialController):
-    def __init__(self, trial_inst: det.Trial, *args: Any, **kwargs: Any) -> None:
-        super().__init__(*args, **kwargs)
+class TrainUnit:
+    def __init__(self, value):
+        self.value = value
+
+    @staticmethod
+    def from_searcher_unit(length: int, unit: core.Unit):
+        if unit == core.Unit.EPOCHS:
+            return Epoch(length)
+        elif unit == core.Unit.RECORDS:
+            return Record(length)
+        elif unit == core.Unit.BATCHES:
+            return Batch(length)
+        else:
+            raise ValueError(f"unrecognized searcher unit {unit}")
+
+    @staticmethod
+    def from_values(batches: Optional[int] = None,
+                    records: Optional[int] = None,
+                    epochs: Optional[int] = None):
+        if sum((batches is not None, records is not None, epochs is not None)) != 1:
+            raise ValueError(f"invalid length: batches={batches} records={records} epochs={epochs}")
+        if batches:
+            return Batch(batches)
+        elif records:
+            return Record(records)
+        elif epochs:
+            return Epoch(epochs)
+
+
+class Epoch(TrainUnit):
+    pass
+
+
+class Batch(TrainUnit):
+    pass
+
+class Record(TrainUnit):
+    pass
+
+class ShouldExit(Exception):
+    """
+    ShouldExit breaks out of the top-level train loop from inside function calls.
+    """
+
+    def __init__(self, skip_exit_checkpoint: bool = False):
+        self.skip_exit_checkpoint = skip_exit_checkpoint
+
+
+class TrialState:
+    def __init__(
+        self,
+        trial_id: int = 0,
+        last_ckpt: int = 0,
+        steps_completed: int = 0,
+        step_id: int = 0,
+        last_val: int = 0,
+        batches_trained: int = None,
```

Comment:
question: why track steps_completed and batches_trained?  It looks like you never actually read state.steps_completed, is that accurate?

Reply:
this is mostly for legacy reasons. it is being read to load from state/checkpoint. i wanted to store batches_trained and epochs_trained because they are more explicit, but the existing code will load state from steps_completed.

Reply:
talked about this: we'll remove steps_completed from the _TrialState